twake/backend/node/src/services/documents/utils.ts

```diff
@@ -105,3 +113,241 @@ const generateAccessToken = (): string => {
+      {
+        company_id: context.company.id,
+        parent_id: item.id,
+      },
+      {},
+      context,
+    );
+
+    Promise.all(
+      children.getEntities().map(async child => {
+        initialSize += await calculateItemSize(child, repository, context);
+      }),
+    );
+
+    return initialSize;
+  }
+
+  return item.size;
+};
+
+/**
+ * Recalculates and updates the Drive item size
+ *
+ * @param {string} id - the item id
+ * @param {Repository<DriveFile>} repository
+ * @param {CompanyExecutionContext} context - the execution context
+ * @returns {Promise<void>}
+ */
+export const updateItemSize = async (
+  id: string,
+  repository: Repository<DriveFile>,
+  context: CompanyExecutionContext,
+): Promise<void> => {
+  if (!id || id === "root" || id === "trash") return;
+
+  const item = await repository.findOne({ id, company_id: context.company.id });
+
+  if (!item) {
+    throw Error("Drive item doesn't exist");
+  }
+
+  item.size = await calculateItemSize(item, repository, context);
+
+  return await repository.save(item);
+};
+
+/**
+ * gets the path for the driveitem
+ *
+ * @param {string} id
+ * @param {Repository<DriveFile>} repository
+ * @param {boolean} ignoreAccess
+ * @param {CompanyExecutionContext} context
+ * @returns
+ */
+export const getPath = async (
+  id: string,
+  repository: Repository<DriveFile>,
+  ignoreAccess?: boolean,
+  context?: CompanyExecutionContext,
+): Promise<DriveFile[]> => {
+  id = id || "root";
+  if (id === "root" || id === "trash") return [];
+
+  const item = await repository.findOne({
+    id,
+    company_id: context.company.id,
+  });
+
+  if (!item || (!checkAccess(id, item, "read", repository, context) && !ignoreAccess)) {
+    return [];
+  }
+
+  return [...(await getPath(item.parent_id, repository, ignoreAccess, context)), item];
+};
+
+/**
+ * checks if access can be granted for the drive item
+ *
+ * @param {string} id
+ * @param {DriveFile | null} item
+ * @param {DriveFileAccessLevel} level
+ * @param {Repository<DriveFile>} repository
+ * @param {CompanyExecutionContext} context
+ * @param {string} token
+ * @returns {Promise<boolean>}
+ */
+export const checkAccess = async (
+  id: string,
+  item: DriveFile | null,
+  level: DriveFileAccessLevel,
+  repository: Repository<DriveFile>,
+  context: CompanyExecutionContext,
+  token?: string,
+): Promise<boolean> => {
+  if (!id || id === "root" || id === "trash") return true;
+
+  try {
+    item =
+      item ||
+      (await repository.findOne({
+        id,
+        company_id: context.company.id,
+      }));
+
+    if (!item) {
+      throw Error("Drive item doesn't exist");
+    }
+
+    if (token) {
+      if (!item.access_info.public.token) {
+        return false;
+      }
+
+      const { token: itemToken, level: itemLevel } = item.access_info.public;
+
+      return hasAccessLevel(itemLevel, level) && itemToken === token;
+    }
+
+    return !!item.access_info.entities.find(async entity => {
```

Comment:
The folder test must comes last, so add some sort here to make sure we end with the channel test and then the folder test.